### PR TITLE
Csv reports next stage

### PR DESCRIPTION
**Pull request by jaleszek**
Implementation of the CSV reports generation for 95% of attributes.
5% of data is not implemented as there is no data structures to pick up this information.
Tested against 11 K set of users/forms, it takes around 60 s per each of them, on development/local.
Added caching association to store the Award form progress per section as it's very memory/time consuming process to use the existing logic.

### CODE REVIEW

File: app/models/form_answer.rb [hunk 1]
```diff
@@ -53,7 +53,10 @@ class FormAnswer < ActiveRecord::Base
     has_one :draft_note, as: :notable, dependent: :destroy
     has_one :company_detail, dependent: :destroy
     has_one :palace_invite, dependent: :destroy
+    has_one :form_answer_progress, dependent: :destroy
```
Comment (matthewford) [inline app/models/form_answer.rb]: Inconsistent indentation detected.

File: app/models/form_answer.rb [hunk 2]
```diff
@@ -53,7 +53,10 @@ class FormAnswer < ActiveRecord::Base
...
 
+    belongs_to :primary_assessor, class_name: "Assessor", foreign_key: :primary_assessor_id
```
Comment (matthewford) [inline app/models/form_answer.rb]: Inconsistent indentation detected.

File: app/models/form_answer.rb [hunk 3]
```diff
@@ -53,7 +53,10 @@ class FormAnswer < ActiveRecord::Base
...
+    belongs_to :secondary_assessor, class_name: "Assessor", foreign_key: :secondary_assessor_id
```
Comment (matthewford) [inline app/models/form_answer.rb]: Inconsistent indentation detected.

File: app/models/reports/all_entries.rb [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+      label: "NomineeEmail",
+      method: :nominee_email
+    }
+  ]
+
+  def initialize(year)
+    @year = year
+  end
+
+  def build
```
Comment (matthewford) [inline app/models/reports/all_entries.rb]: Assignment Branch Condition size for build is too high. [15.52/15]

File: app/models/reports/cases_status_report.rb [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+      label: "SubCategory",
+      method: :sub_category
+    }
+  ]
+
+  def initialize(year)
+    @year = year
+  end
+
+  def build
```
Comment (matthewford) [inline app/models/reports/cases_status_report.rb]: Assignment Branch Condition size for build is too high. [15.52/15]

File: app/models/reports/form_answer.rb [hunk 6]
```diff
@@ -58,20 +217,90 @@ def qae_info_source_other
...
+    bool doc("parent_ultimate_control")
   end
 
-  def last_name
-    obj.user.last_name
+  def mso_outcome_agreed
+    rag @moderated.try(:verdict_rate)
+  end
+
+  def organisation_with_ultimate_control
```
Comment (matthewford) [inline app/models/reports/form_answer.rb]: Duplicate methods `organisation_with_ultimate_control` at lines `224, 232` detected.

File: app/models/reports/form_answer.rb [hunk 7]
```diff
@@ -58,20 +217,90 @@ def qae_info_source_other
...
+      doc "head_email"
+    end
+  end
+
+  def head_title
+    doc("head_of_bussines_title")
+  end
+
+  def head_full_name
+    if  business_form?
```
Comment (matthewford) [inline app/models/reports/form_answer.rb]: Unnecessary spacing detected.

File: app/models/reports/form_answer.rb [hunk 8]
```diff
@@ -99,43 +344,119 @@ def user_creation_date
...
 
+  def innovation?
+    obj.innovation?
+  end
+
   def doc(key)
     obj.document[key]
   end
+
+  def bool var
```
Comment (matthewford) [inline app/models/reports/form_answer.rb]: Use def with parentheses when there are parameters.

File: app/models/reports/form_answer.rb [hunk 9]
```diff
@@ -99,43 +344,119 @@ def user_creation_date
...
+    var ? "Yes" : "No"
+  end
+
+  def rag var
```
Comment (matthewford) [inline app/models/reports/form_answer.rb]: Use def with parentheses when there are parameters.

File: app/models/reports/press_book_list.rb [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+      method: :immediate_parent_country
+    },
+    {
+      label: "QAEAgreedPressNote",
+      method: :qao_agreed_press_note
+    }
+  ]
+
+  def initialize(year)
+    @scope = ::FormAnswer.where(award_year_id: year.id, state: ["awarded", "recommended"]).includes(:user)
```
Comment (matthewford) [inline app/models/reports/press_book_list.rb]: Line is too long. [106/100]

File: app/models/users_import/builder.rb [hunk 11]
```diff
@@ -54,9 +54,13 @@ def self.send_mailing
 
   private
 
+  def log(msg)
+    puts msg unless Rails.env.test?
```
Comment (matthewford) [inline app/models/users_import/builder.rb]: Do not write to stdout. Use Rails' logger if you want to log.

File: app/models/form_answer.rb [hunk 12]
```diff
@@ -220,6 +223,13 @@ def set_award_year
   def set_progress
     form = award_form.decorate(answers: HashWithIndifferentAccess.new(document || {}))
     self.fill_progress = form.progress
+
+    unless new_record?
+      progress = (form_answer_progress || build_form_answer_progress)
+      progress.assign_sections(form)
+      progress.save
+    end
+    true
```
Comment (RuslanHamidullin) [inline app/models/form_answer.rb]: @jaleszek set_progress always will return "true"
```
  def set_progress
    form = award_form.decorate(answers: HashWithIndifferentAccess.new(document || {}))
    self.fill_progress = form.progress

    unless new_record?
      progress = (form_answer_progress || build_form_answer_progress)
      progress.assign_sections(form)
      progress.save
    end
    true # <- HERE
  end
```

Comment (RuslanHamidullin) [inline app/models/form_answer.rb]: Looks like this line is not needed

File: app/models/form_answer_progress.rb [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+    doc = {}
+    (0..5).each do |i|
+      doc["section#{i + 1}"] = f_progress(award_form.steps[i])
+    end
+    self.sections = doc
+  end
+
+  private
+
+  def f_progress(p)
```
Comment (dreamfall) [inline app/models/form_answer_progress.rb]: Please, use more clear variables and method names to improve readability.

File: app/models/reports/form_answer.rb [hunk 14]
```diff
@@ -20,8 +23,164 @@ def call_method(methodname)
     end
   end
 
+  def employees
+    meth = subcategory_suffix("employees")
```
Comment (RuslanHamidullin) [inline app/models/reports/form_answer.rb]: This duplicated 3 times:
```
+    if meth
+      b = doc meth.keys.first
+      doc meth.values.first[b]
+    end
```
in employees, final_year_overseas_sales and final_year_total_sales

File: lib/tasks/form_answers.rake [hunk 15]
```diff
@@ -71,4 +71,22 @@ namespace :form_answers do
       f.save(validate: false)
     end
   end
+
+  desc "Populate the form progresses"
+  # Remove after use - it assumes that progress set up before save
+  task populate_form_answer_progresses: :environment do
```
Comment (RuslanHamidullin) [inline lib/tasks/form_answers.rake]: @jaleszek ping me after merge it, that we need to run this on staging / prod servers

**Comment by RuslanHamidullin** [comment]
@jaleszek looks good :+1: , 2 notes above ^

File: app/models/form_answer.rb (hunk 12 above)
Comment (jaleszek) [inline app/models/form_answer.rb]:
```
  describe '#set_progress', focus: true do
    it "sometimes returns false" do
      f = create(:form_answer)

      # AR::Base save
      # if perform_validation is true validations run.
      # If any of them fail the action is cancelled and save returns false.
      allow_any_instance_of(FormAnswerProgress).to receive(:save).and_return(false)
      expect(f.send(:set_progress)).to eq(false)
    end
  end
```

**Comment by jaleszek** [comment]
@RuslanHamidullin take a look please.

**Pull request merged**